internal/pkg/agent/application/upgrade/manual_rollback.go

```diff
@@ -0,0 +1,258 @@
+		if ttlMarker.Version == rollbackVersion && now.Before(ttlMarker.ValidUntil) {
+			// found a valid target
+			targetInstall = versionedHome
+			targetTTLMarker = ttlMarker
+			break
+		}
+	}
+
+	if targetInstall == "" {
+		return "", "", fmt.Errorf("version %q not listed among the available rollbacks: %w", rollbackVersion, ErrNoRollbacksAvailable)
+	}
+
+	prevAgentParsedVersion, err := version.ParseVersion(targetTTLMarker.Version)
+	if err != nil {
+		return "", "", fmt.Errorf("parsing version of target install %+v: %w", targetInstall, err)
+	}
+
+	// write out a fake upgrade marker to make the upgrade details state happy
+	currentHome := paths.VersionedHome(topDir)
+	relCurVersionedHome, err := filepath.Rel(topDir, currentHome)
+	if err != nil {
+		return "", "", fmt.Errorf("getting current install home path %q relative to top %q: %w", currentHome, topDir, err)
+	}
+	curAgentInstall := agentInstall{
+		parsedVersion: agtversion.GetParsedAgentPackageVersion(),
+		version:       release.VersionWithSnapshot(),
+		hash:          release.Commit(),
+		versionedHome: relCurVersionedHome,
+	}
+
+	prevAgentInstall := agentInstall{
+		parsedVersion: prevAgentParsedVersion,
+		version:       targetTTLMarker.Version,
+		hash:          targetTTLMarker.Hash,
+		versionedHome: targetInstall,
+	}
+
+	upgradeDetails := details.NewDetails(release.VersionWithSnapshot(), details.StateRequested, "" /*action.ID*/)
+	err = markUpgrade(log, paths.DataFrom(topDir), now, curAgentInstall, prevAgentInstall, nil /*action*/, upgradeDetails, nil)
+	if err != nil {
+		return "", "", fmt.Errorf("creating upgrade marker: %w", err)
+	}
+
+	// return watcher executable and versionedHome to rollback to
+	watcherExecutable := watcherHelper.SelectWatcherExecutable(topDir, prevAgentInstall, curAgentInstall)
+	return watcherExecutable, targetInstall, nil
+}
+
+func rollbackUsingUpgradeMarker(ctx context.Context, log *logger.Logger, watcherHelper WatcherHelper, topDir string, now time.Time, version string) (string, string, error) {
+	// read the upgrade marker
+	updateMarker, err := LoadMarker(paths.DataFrom(topDir))
+	if err != nil {
+		return "", "", fmt.Errorf("loading marker: %w", err)
+	}
+
+	if updateMarker == nil {
+		return "", "", ErrNilUpdateMarker
+	}
+
+	// extract the agent installs involved in the upgrade and select the most appropriate watcher executable
+	previous, current, err := extractAgentInstallsFromMarker(updateMarker)
+	if err != nil {
+		return "", "", fmt.Errorf("extracting current and previous install details: %w", err)
+	}
+	watcherExecutable := watcherHelper.SelectWatcherExecutable(topDir, previous, current)
+
+	var selectedRollbackVersionedHome string
+
+	err = withTakeOverWatcher(ctx, log, topDir, watcherHelper, func() error {
+		// read the upgrade marker
+		updateMarker, err = LoadMarker(paths.DataFrom(topDir))
+		if err != nil {
+			return fmt.Errorf("loading marker: %w", err)
+		}
+
+		if updateMarker == nil {
+			return ErrNilUpdateMarker
+		}
```

Comment:
Do you need this code? Isn't `updateMarker` already in scope?  Or is it possible the update marker file could disappear between now and when this closure gets invoked?

Reply:
the difference is that at this point I have the guarantee that no other process is manipulating the upgrade marker, so I reload the most recent state.
The previous upgrade marker loading is there to detect the more recent agent version in order to re-launch the appropriate watcher in case of rollback failure.